File: kotlin-typescript/karakum/plugins/convertArrayInheritance.js
import ts from "typescript";
import * as karakum from "karakum";

export default function (node, context, render) {
    if (
        ts.isExpressionWithTypeArguments(node)
        && ts.isIdentifier(node.expression)
        && (
            node.expression.text === "Array"
            || node.expression.text === "ReadonlyArray"
        )
    ) {
        const typeArguments = node.typeArguments
            ?.map(typeArgument => render(typeArgument))
            ?.filter(Boolean)
            ?.join(", ")

        return `ReadonlyArrayAdapter${karakum.ifPresent(typeArguments, it => `<${it}>`)}`
    }

    return null
}

File: kotlin-typescript/karakum/src/jsMain/kotlin/typescript/karakum/plugins/convertArrayInheritance.kt
package typescript.karakum.plugins

import arrow.core.raise.nullable
import io.github.sgrishchenko.karakum.extension.createPlugin
import io.github.sgrishchenko.karakum.extension.ifPresent
import typescript.asArray
import typescript.isExpressionWithTypeArguments
import typescript.isIdentifier

val convertArrayInheritance = createPlugin { node, _, render ->
    nullable {
        ensure(isExpressionWithTypeArguments(node))

        val expression = node.expression
        ensure(isIdentifier(expression))

        ensure(
            expression.text == "Array"
                    || expression.text == "ReadonlyArray"
        )

        val typeArguments = node.typeArguments?.asArray()
            ?.map { render(it) }
            ?.filter { it.isNotEmpty() }
            ?.joinToString(", ")

        "ReadonlyArrayAdapter${ifPresent(typeArguments) { "<${it}>"}}"
    }
}
